source/mir/ndslice/algorithm.d

```diff
@@ -1884,3 +1885,714 @@ size_t countImpl(alias fun, Slices...)(Slices slices)
     while(!slices[0].empty);
     return ret;
 }
+
+version(unittest)
+{
+    T identity(T)(T x)
+    {
+        return x;
+    }
+}
+
+/++
+The call `eachLower!(fun)(matrix)` applies `fun` to the lower triangle of
+a two-dimensional slice.
+
+The value `k` determines which diagonals will have the function applied:
+For k = 0, the function is also applied to the main diagonal
+For k = 1 (default), only the non-main diagonals below the main diagonal
+will have the function applied.
+For k > 1, fewer diagonals below the main diagonal will have the function
+applied.
+For k < 0, more diagonals above the main diagonal will have the function
+applied.
+
+Params:
+    fun = A function
+    k = adjustment to diagonals (default = 1)
++/
+template eachLower(alias fun, ptrdiff_t k = 1)
+{
+    import mir.functional : naryFun;
+    import mir.ndslice.algorithm : eachImpl;
+    import mir.ndslice.slice : Slice, SliceKind;
+
+    static if (__traits(isSame, naryFun!fun, fun))
+    {
+        void eachLower(SliceKind kind, Iterator)
+                           (Slice!(kind, [2], Iterator) matrix)
+        {
+            immutable(size_t) m = matrix.length!0;
+            immutable(size_t) n = matrix.length!1;
```

Comment:
Oh, lets replace the implementation with single one loop and without `eachUploPair` calls (sorry for wrong recommendation before, found algo issues with it now). We should be careful about code time generation and object code size.

```
x x x x < | 2D eachImpl
x x x x < |
x x x x < || Loop over 1D eachImpl
  x x x < ||
    x x < ||
      x < ||
````

Reply:
I think I have implemented this in the latest commit.